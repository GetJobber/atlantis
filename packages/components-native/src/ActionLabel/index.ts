export { ActionLabel, ActionLabelVariation } from "./ActionLabel";
