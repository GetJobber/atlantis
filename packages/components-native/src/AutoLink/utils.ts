import { EmailMatch, Match, PhoneMatch } from "autolinker";
import { Linking } from "react-native";
import { copyTextToClipboard } from "./clipboard";
import { I18nKeys, useAtlantisI18nValue } from "../hooks/useAtlantisI18n";

function hasPrefix(text: string, prefixes: string[]): boolean {
  return prefixes.some(prefix => text.includes(prefix));
}

export function shouldIgnoreURL(text: string, match: Match): boolean {
  const matchedText = match.getMatchedText().toLocaleLowerCase();
  const urlPrefixes = ["http://", "https://", "www."];
  const ignorePrefixes = ["file://", "ftp://"];
  const previousChar = text.charAt(match.getOffset() - 1);

  if (match.getType() === "url") {
    const hasUrlPrefix = hasPrefix(matchedText, urlPrefixes);
    const hasIgnorePrefix =
      hasPrefix(matchedText, ignorePrefixes) || previousChar === "/";

    return hasIgnorePrefix && !hasUrlPrefix;
  }

  return false;
}

export function getUrl(match: Match, immediateOpen = true): string {
  const matchType = match.getType();

  switch (matchType) {
    case "email":
      return immediateOpen
        ? `mailto:${encodeURIComponent((match as EmailMatch).getEmail())}`
        : (match as EmailMatch).getEmail();
    case "phone":
      return immediateOpen
        ? `tel:${(match as PhoneMatch).getNumber()}`
        : (match as PhoneMatch).getNumber();
    default:
      return match.getAnchorHref();
  }
}

export function onLongPressLink(
  match: Match,
  bottomTabsVisible: boolean,
  t: useAtlantisI18nValue["t"],
): void {
  const linkUrl = getUrl(match, false);

  const toastConfig = {
    message: t(getMessageKey(match)),
    bottomTabsVisible,
  };
  copyTextToClipboard(linkUrl, toastConfig);
}

export function onPressLink(match: Match): void {
  const linkUrl = getUrl(match);
  Linking.openURL(linkUrl);
}

function getMessageKey(match: Match): I18nKeys {
  switch (match.getType()) {
    case "email":
      return "AutoLink.emailCopied";
    case "phone":
      return "AutoLink.phoneCopied";
    default:
      return "AutoLink.urlCopied";
  }
}
