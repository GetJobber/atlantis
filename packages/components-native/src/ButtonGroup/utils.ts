import React, { useCallback } from "react";
import { useIntl } from "react-intl";
import { Alert } from "react-native";
import { messages } from "./messages";
import {
  ButtonGroupActionElement,
  ButtonGroupPrimaryActionElement,
} from "./types";
import {
  ButtonGroupActionProps,
  PrimaryAction,
  SecondaryAction,
} from "./ButtonGroupAction";
import { useAtlantisContext } from "../AtlantisContext/AtlantisContext";

interface UsePreventTapWhenOfflineShape {
  readonly handlePress: (
    callback: ButtonGroupActionProps["onPress"],
  ) => () => void;
}

/**
 * Determine if the onPress should be fired or an alert when the device is
 * online or offline
 */
export function usePreventTapWhenOffline(): UsePreventTapWhenOfflineShape {
  const { formatMessage } = useIntl();
  const { isOnline } = useAtlantisContext();

  const handlePress = useCallback(
    (callback: ButtonGroupActionProps["onPress"]) => {
      if (isOnline) return callback;

      return () =>
        Alert.alert(
          formatMessage(messages.unavailableNetworkTitle),
          formatMessage(messages.unavailableNetworkMessage),
        );
    },
    [formatMessage, isOnline],
  );

  return { handlePress };
}

interface GetActionShape {
  readonly primaryActions: ButtonGroupPrimaryActionElement[];
  readonly secondaryActions: ButtonGroupActionElement[];
}

/**
 * Separate the Primary and Secondary actions into 2 const
 */
export function getActions(
  children: ButtonGroupActionElement | ButtonGroupActionElement[],
): GetActionShape {
  const childArray = React.Children.toArray(
    children,
  ) as ButtonGroupActionElement[];

  let primaryActions = childArray.filter(
    action =>
      /* Checking the component type does not work in dev environments due to
       * wrapper components used for hot-reload
       * However, checking the type name does not work in prod due to code minification
       * Hence 2 different checks here */
      action.type === PrimaryAction || action.type.name === "PrimaryAction",
  ) as ButtonGroupPrimaryActionElement[];

  let secondaryActions = childArray.filter(
    action =>
      action.type === SecondaryAction || action.type.name === "SecondaryAction",
  );

  if (primaryActions.length > 2) {
    secondaryActions = primaryActions.slice(2).concat(secondaryActions);
    primaryActions = primaryActions.slice(0, 2);
  }

  if (primaryActions.length === 0) {
    primaryActions = secondaryActions.slice(0, 1);
    secondaryActions = secondaryActions.slice(1);
  }

  return { primaryActions, secondaryActions };
}
