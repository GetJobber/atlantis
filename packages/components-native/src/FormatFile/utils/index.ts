export { computeA11yLabel } from "./computeA11yLabel";
