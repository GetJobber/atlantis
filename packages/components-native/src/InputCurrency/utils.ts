export function countDecimal(value: number): number {
  const convertedValue = value.toString();

  if (convertedValue.includes(".")) {
    return convertedValue.split(".")[1].length;
  }

  return 0;
}

export function limitInputWholeDigits(
  value: number,
  maxInputLength: number,
): number {
  let convertedValue = value.toString();

  if (convertedValue.length > maxInputLength) {
    convertedValue = convertedValue.slice(0, maxInputLength);

    return parseFloat(convertedValue);
  }

  return value;
}

export function configureDecimal(
  decimalCount: number,
  maxDecimalPlaces: number,
  transformedValue: string,
  decimalPlaces: number,
): number {
  const targetDecimalPlaces = Math.min(
    Math.max(decimalCount, decimalPlaces),
    maxDecimalPlaces,
  );
  const precision = 10 ** targetDecimalPlaces;
  const convertedValue =
    Math.round(parseFloat(transformedValue) * precision) / precision;

  return convertedValue;
}

export function convertToNumber(value: string): string | number {
  const regexValidation = /^[0-9]*$/;

  if (value?.match?.(regexValidation)) {
    return parseFloat(value);
  }

  return value;
}

export const checkLastChar = (stringValue: string): boolean => {
  const lastChar = stringValue[stringValue.length - 1];

  return Boolean(Number(stringValue)) && lastChar !== "0" && lastChar !== ".";
};

export const isValidNumber = (
  numberedValue: string | number | undefined,
): boolean => {
  return (
    typeof numberedValue === "number" &&
    !isNaN(numberedValue) &&
    numberedValue !== 0
  );
};

export const getDecimalNumbers = (
  value: string,
  decimalSeparator: string,
): string => {
  const decimalValue = value.split(".")[1];

  if (!decimalValue) {
    return decimalSeparator;
  }

  return `${decimalSeparator}${decimalValue}`;
};

export const parseGivenInput = (
  value: string | undefined,
  decimalSeparator: string,
): [number, string | undefined, string] => {
  let decimalCount = 0;
  let decimalNumbers = "";
  let wholeIntegerValue = value;

  if (value?.includes(".")) {
    const splittedValue = value?.split(".");
    decimalCount = splittedValue[1].length;
    wholeIntegerValue = splittedValue[0];
    decimalNumbers = getDecimalNumbers(value, decimalSeparator);
  }

  return [decimalCount, wholeIntegerValue, decimalNumbers];
};

export const NUMBER_VALIDATION_REGEX =
  /^[-+]?(([0-9]*\.[0-9]+)|([0-9]+)|([0-9]+(\.?[0-9]+)?e[-+]?[0-9]+))$/;
