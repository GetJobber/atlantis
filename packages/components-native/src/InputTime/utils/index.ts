import { getTimezoneOffset } from "date-fns-tz";

export type MinutesIncrement = 15 | 30 | 60;

/**
 * Rounds up the time by increment.
 * - 15 mins - rounds to the next quarter time of `00:15`, `00:30`, `00:45`,
 *   and `01:00`
 * - 30 mins - rounds to the next half hour be it `00:30` or `01:00`
 * - 60 mins - rounds to the next hour. I.e., `02:01` gets rounded up
 *   to `03:00`.
 */
export function roundUpToNearestMinutes(
  date: Date,
  minutes: MinutesIncrement,
): Date {
  const ms = 1000 * 60 * minutes;

  return new Date(Math.ceil(date.getTime() / ms) * ms);
}

export function getTimeZoneOffsetInMinutes(
  timeZone: string,
  date?: Date,
): number {
  return getTimezoneOffset(timeZone, date) / 1000 / 60;
}
