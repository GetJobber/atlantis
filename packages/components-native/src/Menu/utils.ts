import { LayoutRectangle } from "react-native";
import { styles } from "./Menu.style";

interface ScreenInfo {
  windowHeight: number;
  headerHeight: number;
  windowWidth: number;
}

export function findViewpoint(
  screenInfo: ScreenInfo,
  activatorLayout: LayoutRectangle,
): { [key: string]: number | undefined } {
  const { windowHeight, windowWidth, headerHeight } = screenInfo;
  const pos: { [key: string]: number | undefined } = {};
  const menuWidth = styles.menu.width;
  const windowHalf = (windowHeight - headerHeight) / 2 + headerHeight;
  const menuPositionVertical =
    activatorLayout.y + activatorLayout.height > windowHalf
      ? "menuAbove"
      : "menuBelow";

  const menuPositionHorizontal =
    windowWidth / 2 > activatorLayout.width / 2 + activatorLayout.x
      ? "menuRight"
      : "menuLeft";

  const menuPadding = 36;

  getVerticalPosition(pos, windowHeight, activatorLayout, menuPositionVertical);

  getHorizontalPosition(
    pos,
    activatorLayout,
    windowWidth,
    menuPadding,
    menuWidth,
    menuPositionHorizontal,
  );

  return pos;
}

function getVerticalPosition(
  pos: { [key: string]: number | undefined },
  windowHeight: number,
  activatorLayout: LayoutRectangle,
  menuPositionVertical: string,
) {
  if (menuPositionVertical === "menuAbove") {
    getAbovePosition(pos, windowHeight, activatorLayout);
  }

  if (menuPositionVertical === "menuBelow") {
    getBelowPosition(pos, activatorLayout);
  }
}

function getBelowPosition(
  pos: { [key: string]: number | undefined },
  activatorLayout: LayoutRectangle,
) {
  pos.top = activatorLayout.y + activatorLayout.height;
}

function getAbovePosition(
  pos: { [key: string]: number | undefined },
  windowHeight: number,
  activatorLayout: LayoutRectangle,
) {
  pos.bottom = windowHeight - activatorLayout.y;
}

function getHorizontalPosition(
  pos: { [key: string]: number | undefined },
  activatorLayout: LayoutRectangle,
  windowWidth: number,
  menuPadding: number,
  menuWidth: number,
  menuPositionHorizontal: string,
) {
  if (menuPositionHorizontal === "menuRight") {
    getRightPosition(pos, activatorLayout, windowWidth, menuPadding, menuWidth);
  }

  if (menuPositionHorizontal === "menuLeft") {
    getLeftPosition(pos, activatorLayout, windowWidth, menuPadding, menuWidth);
  }
}

function getLeftPosition(
  pos: { [key: string]: number | undefined },
  activatorLayout: LayoutRectangle,
  windowWidth: number,
  menuHorizontalPadding: number,
  menuWidth: number,
) {
  const overflowLeft =
    windowWidth -
      activatorLayout.x -
      activatorLayout.width +
      activatorLayout.width / 2 -
      menuHorizontalPadding +
      menuWidth >
    windowWidth;

  const overflowRight =
    windowWidth -
      activatorLayout.x -
      activatorLayout.width +
      activatorLayout.width / 2 -
      menuHorizontalPadding <
    0;

  if (overflowLeft) {
    pos.right = undefined;
    pos.left = 0;
  } else if (overflowRight) {
    pos.right = 0;
  } else {
    pos.right = windowWidth - activatorLayout.x - activatorLayout.width;
  }
}

function getRightPosition(
  pos: { [key: string]: number | undefined },
  activatorLayout: LayoutRectangle,
  windowWidth: number,
  menuPadding: number,
  menuWidth: number,
) {
  const overflowRight =
    activatorLayout.x + activatorLayout.width / 2 - menuPadding + menuWidth >
    windowWidth;

  const overflowLeft =
    activatorLayout.x + activatorLayout.width / 2 - menuPadding < 0;

  if (overflowRight) {
    pos.left = undefined;
    pos.right = 0;
  } else if (overflowLeft) {
    pos.left = 0;
  } else {
    pos.left = activatorLayout.x;
  }
}
