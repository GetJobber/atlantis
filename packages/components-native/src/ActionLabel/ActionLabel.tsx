import React from "react";
import { tokens } from "../utils/design";
import type { TextAlign, TextColor } from "../Typography";
import { Typography } from "../Typography";

export type ActionLabelVariation = Extract<
  TextColor,
  "interactive" | "destructive" | "learning" | "subtle" | "onPrimary"
>;

type ActionLabelType = "default" | "cardTitle";

interface ActionLabelProps {
  /**
   * Text to display
   */
  readonly children?: string;

  /**
   * Set the display text to disabled color
   */
  readonly disabled?: boolean;

  /**
   * The text color
   */
  readonly variation?: ActionLabelVariation;

  /**
   * Changes the appearance to match the style of where it's getting used
   */
  readonly type?: ActionLabelType;

  /**
   * Alignment of action label
   */
  readonly align?: TextAlign;
}

export function ActionLabel({
  children,
  variation = "interactive",
  type = "default",
  disabled = false,
  align = "center",
}: ActionLabelProps): React.JSX.Element {
  return (
    <Typography
      color={getColor(variation, disabled)}
      fontFamily="base"
      size="default"
      fontWeight={getFontWeight(type)}
      align={align}
      lineHeight="tight"
      maxFontScaleSize={tokens["typography--fontSize-large"]}
      selectable={false}
    >
      {children}
    </Typography>
  );
}

function getColor(variation: ActionLabelVariation, disabled: boolean) {
  if (disabled) {
    return "disabled";
  }

  if (variation) {
    return variation;
  }

  return "interactive";
}

function getFontWeight(type: ActionLabelType) {
  if (type === "cardTitle") {
    return "bold";
  }

  return "semiBold";
}
