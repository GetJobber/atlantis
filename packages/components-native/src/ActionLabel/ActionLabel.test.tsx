import React, { CSSProperties } from "react";
import { render } from "@testing-library/react-native";
import { ReactTestInstance } from "react-test-renderer";
import { ActionLabel } from "./ActionLabel";
import { tokens } from "../utils/design";

const defaultStyles = {
  fontFamily: "inter-extrabold",
  color: tokens["color-interactive"],
  textAlign: "center",
  fontSize: tokens["typography--fontSize-base"],
  lineHeight: tokens["typography--lineHeight-tight"],
  letterSpacing: tokens["typography--letterSpacing-loose"],
};

describe("ActionLabel", () => {
  it("renders the default action label", () => {
    const text = "Default Action Label";
    const { getByText } = render(<ActionLabel>{text}</ActionLabel>);

    const el = getByText(text);
    expect(el).toBeDefined();
    expect(getStyleObject(el)).toMatchObject(defaultStyles);
  });

  describe("Variations", () => {
    it("renders a destructive variation", () => {
      const text = "Destructive Action Label";
      const { getByText } = render(
        <ActionLabel variation="destructive">{text}</ActionLabel>,
      );

      expect(getStyleObject(getByText(text))).toMatchObject({
        ...defaultStyles,
        color: tokens["color-destructive"],
      });
    });

    it("renders a learning variation", () => {
      const text = "Learning Action Label";
      const { getByText } = render(
        <ActionLabel variation="learning">{text}</ActionLabel>,
      );

      expect(getStyleObject(getByText(text))).toMatchObject({
        ...defaultStyles,
        color: tokens["color-informative"],
      });
    });

    it("renders a subtle variation", () => {
      const text = "Subtle Action Label";
      const { getByText } = render(
        <ActionLabel variation="subtle">{text}</ActionLabel>,
      );

      expect(getStyleObject(getByText(text))).toMatchObject({
        ...defaultStyles,
        color: tokens["color-interactive--subtle"],
      });
    });

    it("renders an onPrimary variation", () => {
      const text = "onPrimary Action Label";
      const { getByText } = render(
        <ActionLabel variation="onPrimary">{text}</ActionLabel>,
      );

      expect(getStyleObject(getByText(text))).toMatchObject({
        ...defaultStyles,
        color: tokens["color-surface"],
      });
    });
  });

  describe("when action label is disabled", () => {
    it("renders text with disabled color, overriding variation", () => {
      const text = "Disabled Action Label";
      const { getByText } = render(
        <ActionLabel disabled variation="destructive">
          {text}
        </ActionLabel>,
      );

      const styles = getStyleObject(getByText(text));
      expect(styles).toMatchObject({
        ...defaultStyles,
        color: tokens["color-disabled"],
      });
      expect(styles).not.toHaveProperty("color", tokens["color-destructive"]);
    });
  });

  describe("when action label is aligned", () => {
    it("renders text with left alignment", () => {
      const text = "Left Aligned Action Label";
      const { getByText } = render(
        <ActionLabel align="start">{text}</ActionLabel>,
      );

      expect(getStyleObject(getByText(text))).toMatchObject({
        ...defaultStyles,
        textAlign: "left",
      });
    });

    it("renders text with right alignment", () => {
      const text = "Right Aligned Action Label";
      const { getByText } = render(
        <ActionLabel align="end">{text}</ActionLabel>,
      );

      expect(getStyleObject(getByText(text))).toMatchObject({
        ...defaultStyles,
        textAlign: "right",
      });
    });
  });
});

function getStyleObject(el: ReactTestInstance) {
  return el.props.style.reduce(
    (mergedStyles: CSSProperties, additionalStyles: CSSProperties) => ({
      ...mergedStyles,
      ...additionalStyles,
    }),
    {},
  );
}
