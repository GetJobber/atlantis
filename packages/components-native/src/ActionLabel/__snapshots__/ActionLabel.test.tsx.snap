// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionLabel supports nested inline content 1`] = `
<Text
  accessibilityRole="text"
  adjustsFontSizeToFit={false}
  allowFontScaling={true}
  maxFontSizeMultiplier={1.125}
  selectable={false}
  selectionColor="hsl(86, 100%, 46%)"
  style={
    [
      {
        "fontFamily": "inter-semibold",
      },
      {
        "color": "hsl(107, 58%, 33%)",
      },
      {
        "textAlign": "center",
      },
      {
        "fontSize": 16,
        "lineHeight": 18,
      },
      {
        "letterSpacing": 0,
      },
    ]
  }
>
  Before 
  <Text
    accessibilityRole="text"
    adjustsFontSizeToFit={false}
    allowFontScaling={true}
    collapsable={false}
    maxFontSizeMultiplier={3.125}
    selectable={true}
    selectionColor="hsl(86, 100%, 46%)"
    style={
      [
        {
          "fontFamily": "inter-regular",
        },
        {
          "color": "hsl(107, 58%, 33%)",
        },
        {
          "textAlign": "left",
        },
        {
          "fontSize": 16,
          "lineHeight": 20,
        },
        {
          "letterSpacing": 0,
        },
      ]
    }
  >
    Inner
  </Text>
   After
</Text>
`;
